src/Storages/MergeTree/ReplicatedMergeTreeCleanupThread.cpp

```diff
@@ -51,39 +54,144 @@ void ReplicatedMergeTreeCleanupThread::run()
         tryLogCurrentException(log, __PRETTY_FUNCTION__);
     }
 
+    UInt64 prev_timestamp = prev_cleanup_timestamp_ms.load(std::memory_order_relaxed);
+    UInt64 now_ms = clock_gettime_ns_adjusted(prev_timestamp * 1'000'000) / 1'000'000;
+
+    /// Do not adjust sleep_ms on the first run after starting the server
+    if (prev_timestamp && storage_settings->cleanup_thread_preferred_points_per_iteration)
+    {
+        /// We don't want to run the task too often when the table was barely changed and there's almost nothing to cleanup.
+        /// But we cannot simply sleep max_cleanup_delay_period (300s) when nothing was cleaned up and cleanup_delay_period (30s)
+        /// when we removed something, because inserting one part per 30s will lead to running cleanup each 30s just to remove one part.
+        /// So we need some interpolation based on preferred batch size.
+        auto expected_cleanup_points = storage_settings->cleanup_thread_preferred_points_per_iteration;
+
+        /// How long should we sleep to remove cleanup_thread_preferred_points_per_iteration on the next iteration?
+        Float32 ratio = cleanup_points / expected_cleanup_points;
+        if (ratio == 0)
+            sleep_ms = storage_settings->max_cleanup_delay_period * 1000;
+        else
+            sleep_ms = static_cast<UInt64>(sleep_ms / ratio);
+
+        if (sleep_ms < storage_settings->cleanup_delay_period * 1000)
+            sleep_ms = storage_settings->cleanup_delay_period * 1000;
+        if (storage_settings->max_cleanup_delay_period * 1000 < sleep_ms)
+            sleep_ms = storage_settings->max_cleanup_delay_period * 1000;
+
+        UInt64 interval_ms = now_ms - prev_timestamp;
+        LOG_TRACE(log, "Scheduling next cleanup after {}ms (points: {}, interval: {}ms, ratio: {}, points per minute: {})",
+                  sleep_ms, cleanup_points, interval_ms, ratio, cleanup_points / interval_ms * 60'000);
+    }
+    prev_cleanup_timestamp_ms.store(now_ms, std::memory_order_relaxed);
+
+    sleep_ms += std::uniform_int_distribution<UInt64>(0, storage_settings->cleanup_delay_period_random_add * 1000)(rng);
```

Comment:
@tavplubix is it expected that `cleanup_delay_period_random_add` applied **after** `max_cleanup_delay_period` ?

Reply:
Yes, it always was like this: https://github.com/ClickHouse/ClickHouse/commit/fa776b93c2f541d95fd20c210acac2b3f733cb47

But I agree that it may be confusing